### PR TITLE
Chore(Source S3): Bump to CDK v5 (release candidate)

### PR DESCRIPTION
**Pull request by aaronsteers**
## What

Builds on:

- #46297

Prereq to:

- #45721 

Adds CDK version bump to v5.

### Fixed Untested Codepaths and Brittle Entrypoint Relationship

Prior to this PR, all of the CAT-related failures were un untested code paths. So, it was a 10-20 minute `airbyte-ci` feedback loop to reproduce issues. This PR adds a number of test structures, and it moves all code logic into test blocks that are on the critical path. The exception is "create()" which requires a catalog input. I put notes in the code that we should try to refactor this so that a source can be created _without_ the catalog input and then passed the catalog after initialization - when it is being passed anyway.

There are two new class methods added on `SourceS3`. The first is `SourceS3.create()` which adds as a Source factory, reducing the complexity of creating a source without introducing any breaking changes to the constructor itself. This was a key point of failure for CAT tests, and the updated code is now in the test pass.

The second new class method is `SourceS3.launch()`. As proposed in the Python grooming session a few weeks back, this change allows Sources to launch them, removing the brittle and unnecessarily brittle relationship between Entrypoint and Connector classes, allowing connectors to launch themselves.

### One-Click CAT Tests.

I added `test_acceptance.py` whose logic should eventually should be pushed into the CDK. This takes a 10-20 minute test loop and makes it a 10 second one. You can re-run any failed test directly, and you can also step-debug through any test. Lastly, we can now easily get test coverage metrics which are specific to the acceptance test and CAT scenarios. Any code not called by CAT scenario can be considered untested for those mapped scenarios - and might actually be dead code. More on this in a coming Loom.

## Can this PR be safely reverted and rolled back?

- [x] YES 💚
- [ ] NO ❌

### CODE REVIEW

**Comment by aaronsteers** [comment]
Currently failing acceptance tests:

https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/aj_source-s3_bump-to-latest-cdk/1727822536/762c666e482a9e1560e72327effce68d041a3c20/source-s3/4.9.0/output.html

> image

**Comment by aaronsteers** [comment]
For context: this PR was spun up because of headaches with breaking changes unrelated to the larger refactoring here:

> * [Feat(Source-S3): Use dataframe processing in place of singleton record operations #44194](https://github.com/airbytehq/airbyte/pull/44194)

And this PR replaces the first attempt here:

- #45199 

I've bumped all of these down to draft status, fwiw. The order of operations once I'm off of oncall rotation is:

1. Get the CDK bumped.
2. Fix the concurrent CDK bug, which otherwise prevents file-based CDK to use concurrency in specific instances.
3. Optionally: Perform additional refactoring, if helpful.

Since we're expecting a decent perf boost from 1 and 2, number 3 is not "unstoppable" on its own.

Added context: The larger refactoring investigation also looked into more pervasive changes to the overall paradigm - specifically eliminating unnecessary record loops when we can delegate down to lower-level APIs. Since the CDK is very record-centric, and since there were plenty of gains to achieve with the version bump, plus just fixing what was broken, we pushed the larger refactor out-of-scope.

Some of this context also shared in Slack. And I've created a new channel [#proj-s3-performance-boosts](https://airbytehq-team.slack.com/archives/C07QM5QTCA2) for focused conversations on this initiative.

File: airbyte-integrations/connectors/source-s3/source_s3/utils.py [hunk 1]
```diff
@@ -52,3 +55,13 @@ def multiprocess_queuer(func: Callable, queue: mp.Queue, *args: Any, **kwargs: A
...
+
+
+def airbyte_message_to_json(message: AirbyteMessage) -> str:
+    """Dump the provided AirbyteMessage to a JSON string."""
+    return orjson.dumps(cast(dict, AirbyteMessageSerializer.dump(message))).decode()
+
+
+def airbyte_message_from_json(message_json: str) -> AirbyteMessage:
+    """Create an AirbyteMessage object from the provided JSON string."""
+    return AirbyteMessageSerializer.load(orjson.loads(message_json))
```
Comment (aaronsteers) [inline airbyte-integrations/connectors/source-s3/source_s3/utils.py:67]: We should put these in the CDK.

**Comment by aaronsteers** [comment]
@clnoll, @natikgadzhi - This is ready for final review. Once approved, we can use it for progressive rollout.

Caveat: I am not up to speed yet with how progressive rollouts should be configured. We may need to apply some changes here - such as using an "rc" version number.

File: airbyte-integrations/connectors/source-s3/pyproject.toml [hunk 2]
```diff
@@ -3,7 +3,7 @@ requires = [ "poetry-core>=1.0.0",]
 build-backend = "poetry.core.masonry.api"
 
 [tool.poetry]
-version = "4.8.5"
+version = "4.9.0"
```
Comment (aaronsteers) [inline airbyte-integrations/connectors/source-s3/pyproject.toml:6]: @clnoll - How should versioning be declared if we are using progressive-rollout for this? Should this be an "rc" version, and are there other changes needed?

File: airbyte-integrations/connectors/source-s3/metadata.yaml [hunk 3]
```diff
@@ -10,7 +10,7 @@ data:
   connectorSubtype: file
   connectorType: source
   definitionId: 69589781-7828-43c5-9f63-8925b1c1ccc2
-  dockerImageTag: 4.8.5
+  dockerImageTag: 4.9.0
```
Comment (alafanechere) [inline airbyte-integrations/connectors/source-s3/metadata.yaml:13]:
```suggestion
  dockerImageTag: 4.9.0-rc.1
```

Comment (alafanechere) [reply airbyte-integrations/connectors/source-s3/metadata.yaml:13]: And set:
```
release:
  isReleaseCandidate: true
```

**Review by alafanechere — COMMENTED** [review]
Just dropping comments if you want to make this version a RC for progressive rollout

File: airbyte-integrations/connectors/source-s3/pyproject.toml (hunk 2 above)
Comment (aaronsteers) [reply airbyte-integrations/connectors/source-s3/pyproject.toml]: Done! ✅

File: airbyte-integrations/connectors/source-s3/metadata.yaml (hunk 3 above)
Comment (aaronsteers) [reply airbyte-integrations/connectors/source-s3/metadata.yaml]: Done. ✅

File: airbyte-integrations/connectors/source-s3/metadata.yaml [hunk 4]
```diff
@@ -29,6 +29,7 @@ data:
       enabled: true
   releaseStage: generally_available
   releases:
+    is_release_candidate: true
```
Comment (natikgadzhi) [inline airbyte-integrations/connectors/source-s3/metadata.yaml:32]:
```suggestion
    isReleaseCandidate: true
```

**Review by natikgadzhi — COMMENTED** [review]
With the `isReleaseCandidate`, as long as CI is green, I'm happy!

**Review by natikgadzhi — APPROVED** [review]

**Pull request merged**